File: src/core/jscop2tvm.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <set>
#include <regex>
#include <stdexcept>
#include <sstream>
#include <algorithm>

// Requires nlohmann/json (e.g., install via package manager or download json.hpp)
#include "json.hpp"

using json = nlohmann::json;

// ---------------------------------------------------------------------------
// Helper functions for string manipulation
// ---------------------------------------------------------------------------

std::string trim(const std::string& s) {
    auto start = s.begin();
    while (start != s.end() && std::isspace(*start)) { start++; }
    auto end = s.end();
    do { end--; } while (std::distance(start, end) > 0 && std::isspace(*end));
    return std::string(start, end + 1);
}

std::vector<std::string> split_comma(const std::string& s) {
    std::vector<std::string> result;
    std::stringstream ss(s);
    std::string item;
    while (std::getline(ss, item, ',')) {
        std::string t = trim(item);
        if (!t.empty()) result.push_back(t);
    }
    return result;
}

std::vector<std::string> split_and(const std::string& s) {
    std::regex and_re(R"(\band\b)");
    std::sregex_token_iterator iter(s.begin(), s.end(), and_re, -1);
    std::sregex_token_iterator end;
    std::vector<std::string> parts;
    for (; iter != end; ++iter) {
        std::string p = trim(*iter);
        if (!p.empty()) parts.push_back(p);
    }
    return parts;
}

std::string replace_identifiers(const std::string& input, const std::map<std::string, std::string>& var_map) {
    std::regex id_re(R"([A-Za-z_]\w*)");
    std::sregex_iterator begin(input.begin(), input.end(), id_re);
    std::sregex_iterator end;
    std::string result;
    size_t last_pos = 0;
    for (auto it = begin; it != end; ++it) {
        std::smatch match = *it;
        result += input.substr(last_pos, match.position() - last_pos);
        std::string tok = match.str();
        if (var_map.count(tok)) {
            result += var_map.at(tok);
        } else {
            result += tok;
        }
        last_pos = match.position() + match.length();
    }
    result += input.substr(last_pos);
    return result;
}

// ---------------------------------------------------------------------------
// Data Structures
// ---------------------------------------------------------------------------

struct Bound {
    std::string var;
    std::string lo;
    std::string hi;
};

struct Access {
    std::string kind;
    std::string array;
    std::vector<std::string> indices;
};

struct Statement {
    std::string name;
    std::vector<std::string> ivars;
    std::map<std::string, Bound> bounds;
    std::vector<std::string> schedule_order;
    std::vector<Access> accesses;
};

struct ArrayInfo {
    std::string name;
    std::string type;
    std::vector<std::string> sizes;
};

struct Scop {
    std::string name;
    std::vector<std::string> params;
    std::vector<ArrayInfo> arrays;
    std::vector<Statement> statements;
};

// ---------------------------------------------------------------------------
// Parsers
// ---------------------------------------------------------------------------

std::pair<std::vector<std::string>, std::string> split_params(const std::string& s) {
    // 1. Try matching the standard format with parameters: [params] -> { body }
    std::regex param_re(R"(^\s*\[([^\]]*)\]\s*->\s*\{([\s\S]*)\}\s*$)");
    std::smatch m;
    if (std::regex_match(s, m, param_re)) {
        return {split_comma(m[1].str()), trim(m[2].str())};
    }
    
    // 2. Try matching the parameterless format: { body }
    std::regex no_param_re(R"(^\s*\{([\s\S]*)\}\s*$)");
    if (std::regex_match(s, m, no_param_re)) {
        // Return an empty parameter list
        return {{}, trim(m[1].str())};
    }
    
    throw std::invalid_argument("Not an ISL set string: " + s);
}

std::pair<std::string, std::vector<std::string>> parse_tuple(const std::string& text) {
    std::regex tuple_re(R"(([A-Za-z_]\w*)?\s*\[([^\]]*)\])");
    std::smatch m;
    std::string t = trim(text);
    if (!std::regex_match(t, m, tuple_re)) {
        throw std::invalid_argument("Expected a tuple like Name[..]: " + text);
    }
    std::string name = m[1].str();
    std::vector<std::string> indices = split_comma(m[2].str());
    return {name, indices};
}

std::string to_exclusive_upper(const std::string& expr, const std::string& op) {
    std::string e = trim(expr);
    return (op == "<") ? e : "(" + e + ") + 1";
}

std::string to_inclusive_lower(const std::string& expr, const std::string& op) {
    std::string e = trim(expr);
    return (op == "<=") ? e : "(" + e + ") + 1";
}

std::map<std::string, Bound> parse_domain_constraints(const std::string& constraints, const std::vector<std::string>& ivars) {
    std::map<std::string, Bound> bounds;
    for (const auto& v : ivars) {
        bounds[v] = {v, "", ""};
    }

    std::vector<std::string> parts = split_and(constraints);
    std::regex chained_re(R"(^\s*(.+?)\s*(<=|<)\s*([A-Za-z_]\w*)\s*(<=|<)\s*(.+?)\s*$)");
    std::regex one_sided_re(R"(^\s*(.+?)\s*(<=|<|>=|>)\s*(.+?)\s*$)");

    for (const auto& p : parts) {
        std::smatch m;
        if (std::regex_match(p, m, chained_re)) {
            std::string v = m[3].str();
            if (bounds.count(v)) {
                bounds[v].lo = to_inclusive_lower(m[1].str(), m[2].str());
                bounds[v].hi = to_exclusive_upper(m[5].str(), m[4].str());
                continue;
            }
        }

        if (std::regex_match(p, m, one_sided_re)) {
            std::string lhs = trim(m[1].str());
            std::string op = m[2].str();
            std::string rhs = trim(m[3].str());

            if (bounds.count(rhs) && !bounds.count(lhs)) {
                std::swap(lhs, rhs);
                if (op == "<") op = ">";
                else if (op == "<=") op = ">=";
                else if (op == ">") op = "<";
                else if (op == ">=") op = "<=";
            }

            if (bounds.count(lhs)) {
                std::string v = lhs;
                if (op == "<" || op == "<=") {
                    bounds[v].hi = to_exclusive_upper(rhs, op);
                } else {
                    std::string inv_op = (op == ">=") ? "<=" : "<";
                    bounds[v].lo = to_inclusive_lower(rhs, inv_op);
                }
            }
        }
    }

    for (const auto& kv : bounds) {
        if (kv.second.lo.empty() || kv.second.hi.empty()) {
            throw std::invalid_argument("Could not derive both bounds for " + kv.first + " from: " + constraints);
        }
    }
    return bounds;
}

std::string strip_memref(const std::string& name) {
    if (name.find("MemRef_") == 0) return name.substr(7);
    return name;
}

// ---------------------------------------------------------------------------
// File Parsing Logic
// ---------------------------------------------------------------------------

Scop parse_jscop(const std::string& path) {
    std::ifstream f(path);
    if (!f.is_open()) throw std::runtime_error("Could not open file: " + path);
    json data;
    f >> data;

    Scop scop;
    scop.name = data.value("name", "scop");
    
    auto ctx_split = split_params(data["context"].get<std::string>());
    scop.params = ctx_split.first;

    if (data.contains("arrays")) {
        for (const auto& arr_json : data["arrays"]) {
            ArrayInfo arr;
            arr.name = arr_json["name"].get<std::string>();
            arr.type = arr_json["type"].get<std::string>();
            for (const auto& s : arr_json["sizes"]) {
                if (s.is_string()) arr.sizes.push_back(s.get<std::string>());
                else arr.sizes.push_back(std::to_string(s.get<int>()));
            }
            scop.arrays.push_back(arr);
        }
    }

    for (const auto& st_json : data["statements"]) {
        Statement stmt;
        
        // Parse Domain
        auto dom_split = split_params(st_json["domain"].get<std::string>());
        std::string body = dom_split.second;
        std::string tup, constraints;
        size_t colon_pos = body.find(':');
        if (colon_pos != std::string::npos) {
            tup = body.substr(0, colon_pos);
            constraints = body.substr(colon_pos + 1);
        } else {
            tup = body;
        }
        
        auto tup_parsed = parse_tuple(tup);
        stmt.name = tup_parsed.first;
        stmt.ivars = tup_parsed.second;
        
        if (trim(constraints).empty()) {
            throw std::invalid_argument("Domain has no constraints");
        }
        stmt.bounds = parse_domain_constraints(constraints, stmt.ivars);

        // Parse Schedule
        auto sched_split = split_params(st_json["schedule"].get<std::string>());
        std::string sched_body = sched_split.second;
        size_t arrow_pos = sched_body.find("->");
        if (arrow_pos == std::string::npos) throw std::invalid_argument("Schedule missing '->'");
        
        auto rhs_tup = parse_tuple(sched_body.substr(arrow_pos + 2));
        for (const std::string& d : rhs_tup.second) {
            std::string d_trim = trim(d);
            if (std::find(stmt.ivars.begin(), stmt.ivars.end(), d_trim) != stmt.ivars.end()) {
                stmt.schedule_order.push_back(d_trim);
            } else {
                std::regex const_re(R"(-?\d+)");
                if (!std::regex_match(d_trim, const_re)) {
                    throw std::invalid_argument("Unsupported schedule dim: " + d_trim);
                }
            }
        }
        for (const auto& v : stmt.ivars) {
            if (std::find(stmt.schedule_order.begin(), stmt.schedule_order.end(), v) == stmt.schedule_order.end()) {
                stmt.schedule_order.push_back(v);
            }
        }

        // Parse Accesses
        for (const auto& acc_json : st_json["accesses"]) {
            Access acc;
            acc.kind = acc_json["kind"].get<std::string>();
            
            auto rel_split = split_params(acc_json["relation"].get<std::string>());
            std::string rel_body = rel_split.second;
            size_t rel_arrow_pos = rel_body.find("->");
            if (rel_arrow_pos == std::string::npos) throw std::invalid_argument("Access missing '->'");
            
            auto acc_rhs_tup = parse_tuple(rel_body.substr(rel_arrow_pos + 2));
            if (acc_rhs_tup.first.empty()) throw std::invalid_argument("Access RHS has no array name");
            
            acc.array = strip_memref(acc_rhs_tup.first);
            acc.indices = acc_rhs_tup.second;
            stmt.accesses.push_back(acc);
        }
        scop.statements.push_back(stmt);
    }
    return scop;
}

// ---------------------------------------------------------------------------
// Codegen Utils
// ---------------------------------------------------------------------------

std::map<std::string, std::string> classify_axes(const Statement& stmt) {
    std::set<std::string> write_vars;
    for (const auto& acc : stmt.accesses) {
        if (acc.kind == "write") {
            for (const auto& idx : acc.indices) {
                std::regex word_re(R"([A-Za-z_]\w*)");
                std::sregex_iterator begin(idx.begin(), idx.end(), word_re), end;
                for (auto it = begin; it != end; ++it) {
                    std::string tok = it->str();
                    if (std::find(stmt.ivars.begin(), stmt.ivars.end(), tok) != stmt.ivars.end()) {
                        write_vars.insert(tok);
                    }
                }
            }
        }
    }
    std::map<std::string, std::string> kinds;
    for (const auto& v : stmt.ivars) {
        kinds[v] = write_vars.count(v) ? "S" : "R";
    }
    return kinds;
}

std::string get_dtype(const std::string& type) {
    if (type == "float") return "float32";
    if (type == "double") return "float64";
    if (type == "int") return "int32";
    if (type == "long") return "int64";
    return type;
}

std::string zero_for(const std::string& dtype) {
    if (dtype.find("float") == 0) return "T." + dtype + "(0.0)";
    return "T." + dtype + "(0)";
}

std::string index_str(const std::vector<std::string>& indices, const std::map<std::string, std::string>& var_map) {
    std::string result;
    for (size_t i = 0; i < indices.size(); ++i) {
        result += replace_identifiers(indices[i], var_map);
        if (i < indices.size() - 1) result += ", ";
    }
    return result;
}

std::string join(const std::vector<std::string>& vec, const std::string& delim) {
    std::string res;
    for (size_t i = 0; i < vec.size(); ++i) {
        res += vec[i];
        if (i < vec.size() - 1) res += delim;
    }
    return res;
}

// ---------------------------------------------------------------------------
// TVM TensorIR Emission
// ---------------------------------------------------------------------------

std::string emit_tvm(const Scop& scop) {
    if (scop.statements.size() != 1) {
        throw std::runtime_error("Only single-statement SCoPs are supported.");
    }
    const Statement& stmt = scop.statements[0];
    auto kinds = classify_axes(stmt);

    const auto& loop_vars = stmt.schedule_order;
    std::map<std::string, std::string> var_map;
    for (const auto& v : stmt.ivars) var_map[v] = "v_" + v;

    std::vector<std::string> extents;
    for (const auto& v : loop_vars) {
        const Bound& b = stmt.bounds.at(v);
        if (trim(b.lo) == "0") extents.push_back(b.hi);
        else extents.push_back("(" + b.hi + ") - (" + b.lo + ")");
    }

    std::vector<Access> reads, writes;
    std::set<std::string> seen_r, seen_w;
    for (const auto& acc : stmt.accesses) {
        std::string key = acc.array + "|" + join(acc.indices, ",");
        if (acc.kind == "read" && seen_r.insert(key).second) reads.push_back(acc);
        else if (acc.kind == "write" && seen_w.insert(key).second) writes.push_back(acc);
    }

    if (writes.size() != 1) throw std::runtime_error("Expected exactly one write access.");
    const Access& out = writes[0];

    std::string out_dtype = "float32";
    for (const auto& arr : scop.arrays) {
        if (arr.name == out.array) {
            out_dtype = get_dtype(arr.type);
            break;
        }
    }

    std::vector<ArrayInfo> sig_arrays;
    for (const auto& arr : scop.arrays) {
        bool is_read = false;
        for (const auto& r : reads) if (r.array == arr.name) is_read = true;
        if (is_read && arr.name != out.array) sig_arrays.push_back(arr);
    }
    for (const auto& arr : scop.arrays) {
        if (arr.name == out.array) { sig_arrays.push_back(arr); break; }
    }

    std::vector<std::string> params_sig_vec;
    for (const auto& arr : sig_arrays) {
        std::vector<std::string> dims;
        for (const auto& s : arr.sizes) dims.push_back((s == "*") ? "N" : s);
        std::string dtype = get_dtype(arr.type);
        params_sig_vec.push_back(arr.name + ": T.Buffer((" + join(dims, ", ") + ",), \"" + dtype + "\")");
    }

    const Access* self_load = nullptr;
    std::vector<Access> rhs_factors;
    for (const auto& r : reads) {
        if (r.array == out.array && r.indices == out.indices) self_load = &r;
        else rhs_factors.push_back(r);
    }

    bool has_reduction = false;
    for (const auto& kv : kinds) { if (kv.second == "R") has_reduction = true; }

    std::string out_idx = index_str(out.indices, var_map);
    std::vector<std::string> rhs_terms;
    for (const auto& r : rhs_factors) {
        rhs_terms.push_back(r.array + "[" + index_str(r.indices, var_map) + "]");
    }
    std::string rhs_expr = rhs_terms.empty() ? zero_for(out_dtype) : join(rhs_terms, " * ");

    std::string s_kinds;
    std::vector<std::string> bound_lhs_vec;
    for (const auto& v : loop_vars) {
        s_kinds += kinds.at(v);
        bound_lhs_vec.push_back(var_map.at(v));
    }

    std::vector<std::string> reads_list_vec;
    for (const auto& r : reads) {
        reads_list_vec.push_back(r.array + "[" + index_str(r.indices, var_map) + "]");
    }

    std::stringstream ss;
    ss << "import tvm\n";
    ss << "from tvm.script import tir as T\n\n";
    ss << "@T.prim_func\n";
    ss << "def " << scop.name << "(" << join(params_sig_vec, ", ") << ") -> None:\n";
    ss << "    for " << join(loop_vars, ", ") << " in T.grid(" << join(extents, ", ") << "):\n";
    ss << "        with T.block(\"" << out.array << "\"):\n";
    ss << "            " << join(bound_lhs_vec, ", ") << " = T.axis.remap(\"" << s_kinds << "\", [" << join(loop_vars, ", ") << "])\n";
    ss << "            T.reads(" << join(reads_list_vec, ", ") << ")\n";
    ss << "            T.writes(" << out.array << "[" << out_idx << "])\n";

    if (has_reduction && self_load) {
        ss << "            with T.init():\n";
        ss << "                " << out.array << "[" << out_idx << "] = " << zero_for(out_dtype) << "\n";
        ss << "            " << out.array << "[" << out_idx << "] = " << out.array << "[" << out_idx << "] + " << rhs_expr << "\n";
    } else {
        ss << "            " << out.array << "[" << out_idx << "] = " << rhs_expr << "\n";
    }

    return ss.str();
}

// ---------------------------------------------------------------------------
// CLI Entry Point
// ---------------------------------------------------------------------------

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "usage: " << argv[0] << " <file.jscop>\n";
        return 1;
    }
    try {
        Scop scop = parse_jscop(argv[1]);
        std::cout << emit_tvm(scop);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
    return 0;
}

File: src/core/jscop2tvm.py
"""
jscop2tvm.py - Convert Polly JSCoP files to TVM TensorIR.

Assumes rectangular iteration domains and that the SCoP is fully
expressible in TVM (affine, perfectly-nested-after-scheduling, etc.).
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Optional


# ---------------------------------------------------------------------------
# Tiny parser for the subset of ISL-set strings Polly emits in JSCoP files.
# ---------------------------------------------------------------------------
#
# We only need to handle the shapes that appear here:
#
#   "[N, M] -> { Stmt[i, j] : 0 <= i < N and 0 <= j < M }"        (domain)
#   "[N]    -> { Stmt[i, j] -> [i, j] }"                          (schedule)
#   "[N]    -> { Stmt[i, j, k] -> MemRef_A[i, k] }"               (access)
#
# Strategy: split off the parameter list, then split on "->" / ":" and
# regex-pluck the pieces. Bounds get parsed into a small dict per index.

_PARAM_RE = re.compile(r"^\s*\[([^\]]*)\]\s*->\s*\{(.*)\}\s*$", re.DOTALL)
_TUPLE_RE = re.compile(r"([A-Za-z_]\w*)?\s*\[([^\]]*)\]")


def _split_params(s: str) -> tuple[list[str], str]:
    """`[N, M] -> { body }`  →  (["N", "M"], "body")."""
    m = _PARAM_RE.match(s)
    if not m:
        raise ValueError(f"Not an ISL set string: {s!r}")
    params_raw, body = m.group(1), m.group(2)
    params = [p.strip() for p in params_raw.split(",") if p.strip()]
    return params, body.strip()


def _parse_tuple(text: str) -> tuple[Optional[str], list[str]]:
    """`Stmt[i, j]`  →  ("Stmt", ["i", "j"]).  `[i, j]` → (None, [...])."""
    m = _TUPLE_RE.match(text.strip())
    if not m:
        raise ValueError(f"Expected a tuple like Name[..]: {text!r}")
    name = m.group(1)
    idx_raw = m.group(2).strip()
    indices = [x.strip() for x in idx_raw.split(",")] if idx_raw else []
    return name, indices


@dataclass
class Bound:
    """Bounds for one iteration variable.  We expect `lo <= v < hi` form."""
    var: str
    lo: str   # an expression string, e.g. "0"
    hi: str   # exclusive upper bound, e.g. "N"


def _parse_domain_constraints(constraints: str, ivars: list[str]) -> dict[str, Bound]:
    """
    Parse the part after the colon in a domain string.

    Polly emits constraints joined with `and`. For rectangular domains each
    induction variable shows up in conjuncts like:
        0 <= i0 < N
        0 <= i0 and i0 <= N - 1
        0 <= i0 <= N - 1
    We merge anything we find per variable into a single `lo <= v < hi`.
    """
    bounds: dict[str, Bound] = {v: Bound(v, lo="", hi="") for v in ivars}
    parts = [p.strip() for p in re.split(r"\band\b", constraints) if p.strip()]

    # Patterns we accept:
    #   <expr> <= v <  <expr>      or   <expr> <  v <= <expr>     (chained)
    #   <expr> <= v <= <expr>
    #   v <= <expr>     /     v < <expr>
    #   <expr> <= v     /     <expr> < v
    chained_re = re.compile(
        r"^\s*(?P<lo>.+?)\s*(?P<op1><=|<)\s*(?P<v>[A-Za-z_]\w*)\s*"
        r"(?P<op2><=|<)\s*(?P<hi>.+?)\s*$"
    )
    one_sided_re = re.compile(
        r"^\s*(?P<lhs>.+?)\s*(?P<op><=|<|>=|>)\s*(?P<rhs>.+?)\s*$"
    )

    def _to_exclusive_upper(expr: str, op: str) -> str:
        # `v < hi`  → exclusive hi as-is
        # `v <= hi` → exclusive hi is `(hi) + 1`
        return expr.strip() if op == "<" else f"({expr.strip()}) + 1"

    def _to_inclusive_lower(expr: str, op: str) -> str:
        # `lo <= v` → lower is `lo`
        # `lo <  v` → lower is `(lo) + 1`
        return expr.strip() if op == "<=" else f"({expr.strip()}) + 1"

    for p in parts:
        m = chained_re.match(p)
        if m and m.group("v") in bounds:
            v = m.group("v")
            bounds[v].lo = _to_inclusive_lower(m.group("lo"), m.group("op1"))
            bounds[v].hi = _to_exclusive_upper(m.group("hi"), m.group("op2"))
            continue

        m = one_sided_re.match(p)
        if not m:
            continue
        lhs, op, rhs = m.group("lhs").strip(), m.group("op"), m.group("rhs").strip()

        # Normalize so the induction variable is on the left.
        if rhs in bounds and lhs not in bounds:
            lhs, rhs = rhs, lhs
            op = {"<": ">", "<=": ">=", ">": "<", ">=": "<="}[op]

        if lhs in bounds:
            v = lhs
            if op in ("<", "<="):
                bounds[v].hi = _to_exclusive_upper(rhs, op)
            else:  # >=, >
                inv_op = "<=" if op == ">=" else "<"
                bounds[v].lo = _to_inclusive_lower(rhs, inv_op)

    for v, b in bounds.items():
        if not b.lo or not b.hi:
            raise ValueError(f"Could not derive both bounds for {v} from: {constraints!r}")
    return bounds


# ---------------------------------------------------------------------------
# IR for a single SCoP statement.
# ---------------------------------------------------------------------------

@dataclass
class Access:
    kind: str           # "read" or "write"
    array: str          # canonical array name (e.g. "C", stripped of MemRef_)
    indices: list[str]  # index expressions (one per dim), as strings


@dataclass
class Statement:
    name: str
    ivars: list[str]              # induction variables in domain order
    bounds: dict[str, Bound]
    schedule_order: list[str]     # ivars permuted into schedule order
    accesses: list[Access]


@dataclass
class Scop:
    name: str
    params: list[str]
    arrays: list[dict]            # raw entries from JSCoP
    statements: list[Statement]


# ---------------------------------------------------------------------------
# Parsing the JSCoP file into the IR above.
# ---------------------------------------------------------------------------

def _strip_memref(name: str) -> str:
    return name[len("MemRef_"):] if name.startswith("MemRef_") else name


def _parse_domain(domain_str: str) -> tuple[str, list[str], dict[str, Bound]]:
    _, body = _split_params(domain_str)
    if ":" in body:
        tup, constraints = body.split(":", 1)
    else:
        tup, constraints = body, ""
    name, ivars = _parse_tuple(tup)
    if not constraints.strip():
        raise ValueError(f"Domain has no constraints: {domain_str!r}")
    bounds = _parse_domain_constraints(constraints, ivars)
    return name, ivars, bounds


def _parse_schedule(sched_str: str, ivars: list[str]) -> list[str]:
    """
    Parse `Stmt[i,j,k] -> [j,i,k]` and return the loop order in terms of the
    original induction variables. We assume each output dim is exactly one
    of the input ivars (true for the simple cases this tool targets).
    """
    _, body = _split_params(sched_str)
    if "->" not in body:
        raise ValueError(f"Schedule missing '->': {sched_str!r}")
    lhs, rhs = body.split("->", 1)
    _, _ = _parse_tuple(lhs)
    _, out_dims = _parse_tuple(rhs)
    order = []
    for d in out_dims:
        d = d.strip()
        if d in ivars:
            order.append(d)
        else:
            # Constant dims (e.g. `0`) are scheduling padding from Polly; skip.
            if not re.fullmatch(r"-?\d+", d):
                raise ValueError(f"Unsupported schedule dim {d!r} in {sched_str!r}")
    # Append any ivars the schedule didn't mention, preserving domain order.
    for v in ivars:
        if v not in order:
            order.append(v)
    return order


def _parse_access(rel_str: str, kind: str) -> Access:
    _, body = _split_params(rel_str)
    if "->" not in body:
        raise ValueError(f"Access missing '->': {rel_str!r}")
    lhs, rhs = body.split("->", 1)
    _, _ = _parse_tuple(lhs)
    arr_name, idx = _parse_tuple(rhs)
    if arr_name is None:
        raise ValueError(f"Access RHS has no array name: {rel_str!r}")
    return Access(kind=kind, array=_strip_memref(arr_name), indices=idx)


def parse_jscop(path: str) -> Scop:
    with open(path) as f:
        data = json.load(f)

    params, _ = _split_params(data["context"])
    statements: list[Statement] = []
    for st in data["statements"]:
        name, ivars, bounds = _parse_domain(st["domain"])
        order = _parse_schedule(st["schedule"], ivars)
        accesses = [_parse_access(a["relation"], a["kind"]) for a in st["accesses"]]
        statements.append(
            Statement(name=name, ivars=ivars, bounds=bounds,
                      schedule_order=order, accesses=accesses)
        )
    return Scop(name=data.get("name", "scop"),
                params=params,
                arrays=data.get("arrays", []),
                statements=statements)


# ---------------------------------------------------------------------------
# Axis classification.
# ---------------------------------------------------------------------------
#
# In TVM, every block axis is either Spatial ("S") or Reduction ("R").
# A clean rule for a single statement: an induction variable is Spatial iff
# it appears in some write access's index list; otherwise it's Reduction.
# (For matmul: i0, i1 appear in writes to C → S; i2 only in reads → R.)

def classify_axes(stmt: Statement) -> dict[str, str]:
    write_vars: set[str] = set()
    for acc in stmt.accesses:
        if acc.kind == "write":
            for idx in acc.indices:
                # Pick out simple identifier references; we expect identity
                # indexing for rectangular SCoPs.
                for tok in re.findall(r"[A-Za-z_]\w*", idx):
                    if tok in stmt.ivars:
                        write_vars.add(tok)
    return {v: ("S" if v in write_vars else "R") for v in stmt.ivars}


# ---------------------------------------------------------------------------
# TVM TensorIR codegen.
# ---------------------------------------------------------------------------

_DTYPE_MAP = {"float": "float32", "double": "float64",
              "int": "int32", "long": "int64"}


def _array_decl(arr: dict) -> str:
    sizes = arr["sizes"]
    # Polly often marks the leading dimension as "*" (unknown).
    # For our targeted cases we treat that as the parameter (e.g. N).
    dims = []
    for s in sizes:
        dims.append("N" if s == "*" else str(s))
    dtype = _DTYPE_MAP.get(arr["type"], arr["type"])
    return f"{arr['name']}: T.Buffer(({', '.join(dims)},), \"{dtype}\")"


def _zero_for(dtype: str) -> str:
    if dtype.startswith("float"):
        return f"T.{dtype}(0.0)"
    return f"T.{dtype}(0)"


def _index_str(indices: list[str], var_map: dict[str, str]) -> str:
    """Rewrite domain ivars (`i0`) into block-bound vars (`v_i0`)."""
    def repl(m):
        tok = m.group(0)
        return var_map.get(tok, tok)
    return ", ".join(re.sub(r"[A-Za-z_]\w*", repl, idx) for idx in indices)


def emit_tvm(scop: Scop) -> str:
    if len(scop.statements) != 1:
        # The hint scopes us to single-statement SCoPs; punt cleanly otherwise.
        raise NotImplementedError("Only single-statement SCoPs are supported.")
    stmt = scop.statements[0]
    kinds = classify_axes(stmt)

    # Loop variables in schedule order, plus a remap to TVM's `v_*` block vars.
    loop_vars = stmt.schedule_order
    var_map = {v: f"v_{v}" for v in stmt.ivars}

    # Loop extents (use the bounds' exclusive upper - lower; for rectangular
    # SCoPs starting at 0 this is just `hi`).
    extents = []
    for v in loop_vars:
        b = stmt.bounds[v]
        if b.lo.strip() == "0":
            extents.append(b.hi)
        else:
            extents.append(f"({b.hi}) - ({b.lo})")

    # Determine reads/writes (after dedup) and the array list.
    reads, writes = [], []
    seen_r, seen_w = set(), set()
    for acc in stmt.accesses:
        key = (acc.array, tuple(acc.indices))
        if acc.kind == "read" and key not in seen_r:
            seen_r.add(key); reads.append(acc)
        elif acc.kind == "write" and key not in seen_w:
            seen_w.add(key); writes.append(acc)

    if len(writes) != 1:
        raise NotImplementedError("Expected exactly one write access.")
    out = writes[0]

    # Element dtype of the output array (for the init expression).
    out_dtype = "float32"
    for arr in scop.arrays:
        if arr["name"] == out.array:
            out_dtype = _DTYPE_MAP.get(arr["type"], arr["type"])
            break

    # Build the function signature in TVM order: inputs (read-only) then outputs.
    read_arrays = {a.array for a in reads} - {out.array}
    sig_arrays = [a for a in scop.arrays if a["name"] in read_arrays]
    sig_arrays.append(next(a for a in scop.arrays if a["name"] == out.array))
    params_sig = ", ".join(_array_decl(a) for a in sig_arrays)

    # Heuristic: the body is `out += prod(reads other than self-load)`.
    # The self-load of `out` (read with the same indices as the write) signals
    # an accumulation; we omit it from the RHS and emit a `T.init()`.
    self_load = next(
        (r for r in reads if r.array == out.array and r.indices == out.indices),
        None,
    )
    rhs_factors = [r for r in reads if r is not self_load]
    has_reduction = any(k == "R" for k in kinds.values())

    # Indices, rendered with v_-prefixed variables.
    out_idx = _index_str(out.indices, var_map)
    rhs_terms = [f"{r.array}[{_index_str(r.indices, var_map)}]" for r in rhs_factors]
    rhs_expr = " * ".join(rhs_terms) if rhs_terms else _zero_for(out_dtype)

    # Assemble the TensorIR string.
    s_kinds = "".join(kinds[v] for v in loop_vars)
    grid_args = ", ".join(extents)
    loop_idx = ", ".join(loop_vars)
    bound_lhs = ", ".join(var_map[v] for v in loop_vars)

    reads_list = ", ".join(
        f"{r.array}[{_index_str(r.indices, var_map)}]" for r in reads
    )
    writes_list = f"{out.array}[{out_idx}]"

    lines = [
        "import tvm",
        "from tvm.script import tir as T",
        "",
        "@T.prim_func",
        f"def {scop.name}({params_sig}) -> None:",
        f"    for {loop_idx} in T.grid({grid_args}):",
        f'        with T.block("{out.array}"):',
        f'            {bound_lhs} = T.axis.remap("{s_kinds}", [{loop_idx}])',
        f"            T.reads({reads_list})",
        f"            T.writes({writes_list})",
    ]
    if has_reduction and self_load is not None:
        lines += [
            "            with T.init():",
            f"                {out.array}[{out_idx}] = {_zero_for(out_dtype)}",
            f"            {out.array}[{out_idx}] = "
            f"{out.array}[{out_idx}] + {rhs_expr}",
        ]
    else:
        lines.append(f"            {out.array}[{out_idx}] = {rhs_expr}")
    return "\n".join(lines) + "\n"


# ---------------------------------------------------------------------------
# CLI entry point.
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("usage: python jscop2tvm.py <file.jscop>", file=sys.stderr)
        sys.exit(1)
    scop = parse_jscop(sys.argv[1])
    print(emit_tvm(scop))
